### ISSUE TITLE
How to help?

### ISSUE THREAD

**Issue by c-smile** [c-smile/sciter-js-sdk#10]
Tasks that I think make sense to do at this moment:

#### Idea 

Idea is to wrap ffmpeg lib as video source behavior for Sciter. The code can be based on [behvaior_video_generator](https://github.com/c-smile/sciter-sdk/blob/master/include/behaviors/behavior_video_generator.cpp) and [How to Write a Video Player in Less Than 1000 Lines](http://dranger.com/ffmpeg/) tutorial.

#### Output

Ideally implementation should  support as static linkage as a dll (sciter-ffmpeg.dll). Pretty much in the same way as current [SQLite component](https://github.com/c-smile/sciter-js-sdk/tree/main/sqlite) that supports dual linkage. 

#### Goals

This project will help to establish mechanism for loadable native behaviors.
```CSS
video {
   behavior: ffmpeg-streams "sciter-ffmpeg.dll"
}
```
It is not clear yet what such behaviors will need from Sciter side (other that passing Sciter API reference as it is).

#### Goals

Ideally implementation should be "promise capable" - so the project will help to establish API for native promises.

### VueJS and ReactJS 

Makes sense to try to port any of these.

#### Goals

These will help: 

*  to discover DOM features that Sciter definitely need to support to run Vue or React applications;
*  to decide what to implement in JS in copmptibility-layer.js and what needs to be implemented natively;

### Any popular Charts library

The one that use `<canvas>` for rendering. 

#### Goals

To test and improve my current quite basic `<canvas>` implementation.

### Maps 

[Leaflet](https://leafletjs.com/) anyone?
 
Please let me know if you want to take any one of these or have other ideas.

#### Idea 

That's brand new component that needs to be designed from scratch. And I will help as much as I can on this. 

Seems like many of us need a component that can be built-in into applications.  
Something along the lines of Slack, Skype, etc: 

![chat message tape](https://terrainformatica.com/wp-content/uploads/2019/09/chat3-e1569252441546.png)

#### Goals

This will end up with "virtual tape" - tape of messages. We need to decide if script is enough for that or do I need to add native support for virtual lists.

**Comment by jsprog** [comment]
> Scripting frameworks and components
VueJS and ReactJS
Makes sense to try to port any of these.

I don't know if my request is too heavy for sciter.js, but the future is there. At least, consider adding the features not needed by the app through a dynamic library that will be used solely for development purposes.

I recommend that you focus on the use of [https://github.com/vitejs/vite](https://github.com/vitejs/vite) and [https://www.snowpack.dev/](https://www.snowpack.dev/) whenever possible. Reactive frameworks are moving away from bundlers to Native ES modules during development. The bunlders are staying there to focus solely on bundling production builds. This modern approach is tested to be much lighter for development (creation for projects, neat package.json, starting development Server and HMR).

**[Vite:](https://github.com/vitejs/vite)**
it has templates for creating Vue, react and preact projects with support for HMR out of the box. For production builds, it delegate the bundling work to rollup.
For svelte users, there is a plugin [https://madewithsvelte.com/vite-plugin-svelte](https://madewithsvelte.com/vite-plugin-svelte) and I managed to use it successfully with HMR support (of course, with a web browser).

**[Snowpack:](https://www.snowpack.dev/)**
You gain similar experience as vite. many templates are available out of the box (React, Preact, Svelte, Vue, etc...) with support for HMR. Webpack is used for bundling production builds but rollup is joining soon. If you scroll down the main page, you'll see a graph explaining the differences between the older and the newer approaches.

**Requirements for vite and snowpack**
- ESM Modules: ES2020 is supported by quickjs
- ESM Hot module replacement: [https://modern-web.dev/docs/dev-server/plugins/hmr/](https://modern-web.dev/docs/dev-server/plugins/hmr/): import.meta is available (empty object). Not sure if vite or snowpack will fill this empty object.
- Loading app from HTTP server
- WebSocket

**Comment by c-smile** [comment]
> I don't know if my request is too heavy for sciter.js, but the future is there. At least, consider adding the features not needed by the app through a dynamic library that will be used solely for development purposes.

Could you elaborate more on that?  In particular purpose "a dynamic library that will be used solely for development purposes".
What is that and why do we need it?

> Vite & Co,

Not sure I understand how Sciter.JS is related to all that. And what is so difficult in hot reload?

> Loading app from HTTP server

Are we talking about desktop UI here?

**Comment by jsprog** [comment]
> Could you elaborate more on that? In particular purpose "a dynamic library that will be used solely for development purposes".
> What is that and why do we need it?

Adding some features may improve the developer's experience, but they may not be needed by a production app. You may have some reservations about a possible negative impact on binary size, and that's why I'm suggesting a work around for this if any.

> Not sure I understand how Sciter.JS is related to all that. And what is so difficult in hot reload?

> Are we talking about desktop UI here?

You already said '**Makes sense to try to port any of these.**', and that's whey I wanted to share my thoughts. Usually, developers depend on ready made templates to kick starting new reactive projects. **Vite** and **Snowpack** are the new school tools to achieve this and Sciter.js is already having supprort for modern javascript syntax. They're just exploiting the web standard, and by focusing on their requirements from the browser side during development, we should be able to use them with sciter.js:

- Loading the app and ES modules from server during development:
  - scapp http://localhost:port
  - import module from './module.js'. The module may come either from local files or server depending how we started the app.

- For hot reload to work, Sciter.js should be able to communicate with the development server using [WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket), and it should be able to replace them. In fact, I haven't done any test with the replacement and it's too early for this unless if the support for WebSocket landed.

In fact, I'm planning to give it some efforts, but let's seek stability first (native modules, parent.removeChild, WebSocket, etc...)

**Comment by intrnl** [comment]
At the moment, Preact breaks because it's trying to check the length of `childNodes`, which doesn't exist yet.

**Comment by intrnl** [comment]
React breaks due to ReactDOM checking [whether or not an element is an instance of `window.HTMLIFrameElement`](https://github.com/facebook/react/blob/v17.0.0/packages/react-dom/src/client/ReactInputSelection.js#L62)

**Comment by intrnl** [comment]
Svelte is able to render a click counter (a button that says how many times it's been clicked) successfully, but clicking the button does nothing because it's trying to change the count value by [mutating `data` of the text node](https://github.com/sveltejs/svelte/blob/v3.29.7/src/runtime/internal/dom.ts#L196).

**Comment by jsprog** [comment]
Please don't throw everything here. I can confirm that your first post is due to `element.childNodes` = `undefined` and better to create a new issue for this instead of messing up this topic. not only this but also `document.createNodeIterator` that's used by malinajs is undefined

For your problem with svelte, just change text.data to text.nodeValue using @rollup/plugin-replace and it's already explained by me in a different issue created by another person

Since you're the one who discovered the problem about `element.childNodes`, I can't claim this for myself and can create a new issue for this.

**Comment by intrnl** [comment]
Yeah I realize, my bad.

**Comment by jsprog** [comment]
In fact, I registered in the forums recently to share my ideas because they aren't issues, but my comments aren't landing there.

**Comment by c-smile** [comment]
To @jsprog : it does not matter at this stage where feature requests will appear. 

But all this should really be more organized, so forum is better, I think. But it could be also Wiki page here.

At this phase (this November) I am aiming on two goals:

1. implementation of minimal set of DOM APIs that is needed for popular JS frameworks so I appreciate list of mandatory methods.
2. to make inspector operational: console, DOM tree, style list, JS debugger.

**Comment by c-smile** [comment]
> * Svelte is able to render a click counter (a button that shows how many times it's been clicked) successfully, but clicking the button does nothing because it's trying to change the count value through [`TextNode#data`](https://github.com/sveltejs/svelte/blob/v3.29.7/src/runtime/internal/dom.ts#L196)

I cannot find any mentioning about `data` property anywhere in specs. Any idea what does it mean?

**Comment by jsprog** [comment]
.data property is valid for 'text nodes' and 'comment nodes' and the information about it is quite limited. just use google and type 'textNode.data vs'

[https://stackoverflow.com/questions/12286975/when-working-with-text-nodes-should-i-use-the-data-nodevalue-textcontent](https://stackoverflow.com/questions/12286975/when-working-with-text-nodes-should-i-use-the-data-nodevalue-textcontent)

svelte is updating state by using `text.data = data;` and the text here is a textNode. To get arround this, I had to replace it using rollup.

According to my last test with malina.js (think of it as a deviation from svelte with support for other features) I could confirm that they're using `.nodeValue`

**Comment by jsprog** [comment]
@c-smile: Anyway to reload the window (like F5) from javascript. I needed this since 2 days to proceed with tests, to later share back my full work.

**Comment by intrnl** [comment]
The `data` property comes from [`CharacterData`](https://developer.mozilla.org/en/docs/Web/API/CharacterData)

**Comment by jsprog** [comment]
I already started my first git repository [jsprog/sciterjs-svelte3-template](https://github.com/jsprog/sciterjs-svelte3-template) to share with others my work about bringing the default svelte template to sciter.js. No content at the moment, but It should be there very soon.

**By doing this, I'm trying to:**
- Help others benefit from existing work to kick start their svelte projects.
- Help others to understand how I'm tracking the missing requirements, and by time they should be able to do similar work by them selves or to be more effective when recommending some features for sciter.js
  - Give a man a fish, you feed him for a day.
  - Teach him to fish, and you feed him for life

**Comment by jsprog** [comment]
to @c-smile: We need an alternative to `Location.reload()` and `Window.this.reload()` isn't available yet

Is it possible to achieve this with javascript only

**Comment by c-smile** [comment]
See: https://github.com/c-smile/sciter-js-sdk/blob/main/samples/debug-peer.js#L65

**Comment by jsprog** [comment]
Thank you, I succeeded with live reload with minimal changes (svelte). It's not fully perfect, but It's working and It should help developers coding faster.
* create a new file `livereload_fix.js` with content from below
* import it from `main.js`
* for production builds, we don't need to import the fix. That's why, we need rollup replacement.
* we need rollup replacement again for other 2 patches:
  * "l.getElementsByTagName('head')[0].appendChild(r)": "document.head.appendChild(r)"
  * "r.id = 'livereloadscript'": "Object.defineProperty(r, 'id', { value: 'livereloadscript' })"

```js
// top of main.js

import './livereload_fix'
```


```js
// livereload_fix.js

// note-1: This isn't final and expect it to change with improvements and some fixes
// note-2: The port is hardcoded to the default value used by livereload (35729)
// note-3: Unwanted logs after closing the app (only if you used 'new WebSocket(uri)' at least once)
// note-4: Handshaking with the dev server isn't implemented yet.
//    - issue: the reason behind this is that the app will crash with "Segmentation fault (core dumped)"
//        if it tried to connect to a disconnected websocket server
// note-5: I'm not sending the 'hello' command on connection 
//        (ws.send(JSON.stringify({ command, 'hello', ...other-params }))
//        - doing this will case the server to send a reply, that I'm trying to avoid this time,
//            because I'm only receiving empty messages
// note-6: I'm only receiving empty messages from the development server,
//    and I'm debouncing them to trigger the reload
//    - likely I'm not sending the 'hello' command with the correct parameters (not sure)
// note-7: the script injection for livereload-js is kept as it is for these reasons:
//    - To keep compatibility with browsers
//    - It has no effect to Sciter.js

import debounce from 'lodash/debounce'

const debounce_duration = 200
const reconnection_timeout = 1000
const ws_uri = `ws://localhost:35729/livereload`

if (Window.this) {
    let ws

    const debounce_then_reload = debounce(() => {
        document.dispatchEvent(new Event("reloaddocument", {bubbles:true}),true);        
    }, debounce_duration)

    function connect () {
        ws = new WebSocket(ws_uri)
        
        ws.onopen = () => {}
 
        // just crash the app if the connection with server is closed.
        //    Better than keeping it to not receive the updates from a restarted server
        ws.onclose = () => { setTimeout(() => { connect() }, reconnection_timeout) }

        ws.onmessage = () => { debounce_then_reload() }

        ws.onerror = (err) => {}
    }
   
    connect()
}
```


```js
// rollup.config.js

production && replace({
    delimiters: ['', ''],
    // we don't need livereload_fix for production builds
    "import './livereload_fix'": ""
}),

replace({
    delimiters: ['', ''],
    // fix: use alternative supported syntax
    "l.getElementsByTagName('head')[0].appendChild(r)": "document.head.appendChild(r)",
    // fix: no setter for property
    "r.id = 'livereloadscript'": "Object.defineProperty(r, 'id', { value: 'livereloadscript' })"
}),

```

```js
// example App.svelte

<script>
    let count = 0
    setInterval(() => {
	count++
	console.log('count:', count)
    }, 1000)
</script>

<div>Count: {count}</div>

```

@c-smile: The next steps are:
* Fixing WebSocket in order to improve `livereload_fix.js`. Likely, I'll open a new issue after some extensive tests with WebSocket.
* Access to routes: we need this for:
  * To preserve navigation after reload (like browsers)
  * It could be done with custom code and persistence, but we have to stick with the routing notion that's to be implemented by Sciter.js

**Comment by c-smile** [comment]
What's the problem with WebSockets?

As of preserving navigation, I will add `data` property. It's content will survive reloading.  Is this what you are looking for?

**Comment by jsprog** [comment]
> What's the problem with WebSockets?

I can't say too much about this without some extensive tests. Sorry about this, I had to spend my day tracking some hard to recreate CSS bugs, and I'll try report about them within 24 hours).

> As of preserving navigation, I will add data property. It's content will survive reloading. Is this what you are looking for?

I think that you understand better than me. I suggested to keep navigation state across reloads and you're suggesting something that would help with:
- preserving navigation (route)
- preserving login state
- preserving other state passed through navigation
- preserving other state like sessionStorage or a simple embedded key/value store, with the difference that we're storing objects rather than strings or binaries.
